service/controller/v3/resource/vpngateway/desired.go

```diff
@@ -0,0 +1,125 @@
+)
+
+// GetDesiredState return desired vpn gateway connections.
+func (r *Resource) GetDesiredState(ctx context.Context, azureConfig interface{}) (interface{}, error) {
+	customObject, err := key.ToCustomObject(azureConfig)
+	if err != nil {
+		return connections{}, microerror.Mask(err)
+	}
+
+	var (
+		hostVPNGateway  *network.VirtualNetworkGateway
+		guestVPNGateway *network.VirtualNetworkGateway
+	)
+	// Do not check for vpn gateway when deleting.
+	// As we do not require guest cluster vpn gateway to be ready in order to
+	// delete connection from host cluster vpn gateway.
+	if !key.IsDeleted(customObject) {
+		// In order to make vpn gateway connection work we need 2 vpn gateway.
+		// One on the host cluster and one on the guest cluster.
+		// Here we check for vpn gateways readiness.
+		// In case one of the vpn gateway is not ready we cancel the resource
+		// and try again on the next resync period.
+		{
+
+			resourceGroup := key.ResourceGroupName(customObject)
+			vpnGatewayName := key.VPNGatewayName(customObject)
+
+			guestVPNGateway, err = r.getGuestVirtualNetworkGateway(ctx, resourceGroup, vpnGatewayName)
+			if IsVPNGatewayNotFound(err) {
+				r.logger.LogCtx(ctx, "level", "debug", "message", "guest vpn gateway was not found")
+				resourcecanceledcontext.SetCanceled(ctx)
+				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+				return connections{}, nil
+			} else if err != nil {
+				return connections{}, microerror.Mask(err)
+			}
+
+			if provisioningState := *guestVPNGateway.ProvisioningState; provisioningState != "Succeeded" {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest vpn gateway is in state '%s'", provisioningState))
+				resourcecanceledcontext.SetCanceled(ctx)
+				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+				return connections{}, nil
+			}
+		}
+
+		{
+			resourceGroup := r.azure.HostCluster.ResourceGroup
+			vpnGatewayName := r.azure.HostCluster.VirtualNetworkGateway
+
+			hostVPNGateway, err = r.getHostVirtualNetworkGateway(ctx, resourceGroup, vpnGatewayName)
+			if IsVPNGatewayNotFound(err) {
+				r.logger.LogCtx(ctx, "level", "debug", "message", "host vpn gateway was not found")
+				resourcecanceledcontext.SetCanceled(ctx)
+				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+				return connections{}, nil
+			} else if err != nil {
+				return connections{}, microerror.Mask(err)
+			}
+
+			if provisioningState := *hostVPNGateway.ProvisioningState; provisioningState != "Succeeded" {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("host vpn gateway is in state '%s'", provisioningState))
+				resourcecanceledcontext.SetCanceled(ctx)
+				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+				return connections{}, nil
+			}
+		}
+	}
```

Comment:
Isn't GetCurrentState better place for that? We shouldn't exercise the Azure API anymore if those two guys aren't ready yet.

Reply:
Yes that was my initial design. But connection need reference to gateway objects, so it makes more sense to have them here in the end.